net: lib: lwm2m: Fix bool value not sent correctly

sarchey opened the pull request:
The GCC minsizerel (-Os) optimizes out the value conversion to a real bool (0 or 1).

koalatux commented:
FYI, here you can see the difference:

- [with `bool` parameter](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:1,endLineNumber:8,positionColumn:1,positionLineNumber:8,selectionStartColumn:1,selectionStartLineNumber:8,startColumn:1,startLineNumber:8),source:'%23include+%3Cstdbool.h%3E%0A%23include+%3Cstdint.h%3E%0A%0Auint8_t+set_bool(bool+value)+%7B%0A++++uint8_t+ret+%3D+value+!!%3D+0+%3F+1+:+0%3B%0A++++return+ret%3B%0A%7D%0A'),l:'5',n:'1',o:'C+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:carmug1420,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-Os',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+ARM+GCC+14.2.0+(unknown-eabi)+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
- [with `uint8_t` parameter](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:1,endLineNumber:8,positionColumn:1,positionLineNumber:8,selectionStartColumn:1,selectionStartLineNumber:8,startColumn:1,startLineNumber:8),source:'%23include+%3Cstdbool.h%3E%0A%23include+%3Cstdint.h%3E%0A%0Auint8_t+set_bool(uint8_t+value)+%7B%0A++++uint8_t+ret+%3D+value+!!%3D+0+%3F+1+:+0%3B%0A++++return+ret%3B%0A%7D%0A'),l:'5',n:'1',o:'C+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:carmug1420,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-Os',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+ARM+GCC+14.2.0+(unknown-eabi)+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

rlubos reviewed (commented):
Please elaborate, does the compiler optimization cause any trouble for you? I start to wonder if that conversion we have is really needed, as bool in C seems to guarantee to be only 0 or 1. 

We shouldn't come lightly to API changes, especially that accepting `bool` type in `set_bool` functions seems quite reasonable.

sarchey commented:
> Please elaborate, does the compiler optimization cause any trouble for you? I start to wonder if that conversion we have is really needed, as bool in C seems to guarantee to be only 0 or 1.
> 
> We shouldn't come lightly to API changes, especially that accepting `bool` type in `set_bool` functions seems quite reasonable.

I discovered a case where a bool value was set with memcpy and got the value 5. Since the GCC compiler assumes that a bool value is only true or false, it optimised the existing check that sets the value to 0 or 1 and then unfortunately a wrong value is sent. This could lead to an error in the backend when this lwm2m message was parsed.

Since the behaviour changes with different compiler options (-O0 works), I suggest to fix this. Either by changing the parameter type or by removing the existing bool-to-0/1 conversion so that the outcome is always the same.

jukkar commented:
> ```c
> int lwm2m_set_bool(const struct lwm2m_obj_path *path, bool value) 
> ```

As the function is about setting a bool value, it would kind of make sense to have a bool parameter like there is atm instead of having a `uint8_t`

jukkar commented:
Can't we do a cast like this?

```c
int lwm2m_set_bool(const struct lwm2m_obj_path *path, bool value) 
{
      uint8_t temp; 

      compiler_barrier(); 

      temp = (uint8_t)value;
      
      return lwm2m_engine_set(path, &temp, 1); 
}     
```

SeppoTakalo commented:
> I discovered a case where a bool value was set with memcpy and got the value 5.

Did this happen in the engine or in the application code?
Because, for me, this sounds like there is a bug that needs to be fixed instead of changing the API to prevent it causing other problems.

sarchey committed:
net: lib: lwm2m: Fix bool value not sent correctly

Change the value conversion from a bool to 0/1 so that the GCC does not
optimize it out with CONFIG_SIZE_OPTIMIZATIONS=y and minsizerel option.

Signed-off-by: Sebastian Schoch <sebastian.schoch@husqvarnagroup.com>

sarchey commented:
> > I discovered a case where a bool value was set with memcpy and got the value 5.
> 
> Did this happen in the engine or in the application code? Because, for me, this sounds like there is a bug that needs to be fixed instead of changing the API to prevent it causing other problems.

This is an bug on the application side and has to be fixed there. However because there is already a "safety" conversion in `lwm2m_set_bool` I prefer to have the same output with every optimization config (CONFIG_DEBUG_OPTIMIZATIONS and CONFIG_SIZE_OPTIMIZATIONS).
 
Since changing the API does not seem to be a preferred solution, I have now replaced the conversion to make it ‘optimisation-proof’.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_registry.c:
```diff
@@ -778,7 +778,11 @@ int lwm2m_set_s64(const struct lwm2m_obj_path *path, int64_t value)
 int lwm2m_set_bool(const struct lwm2m_obj_path *path, bool value)
 {
-	uint8_t temp = (value != 0 ? 1 : 0);
+	uint8_t temp = 0;
```
Empty line is needed after variable definition.
Also, please add a short comment in the code why is this needed.

pdgendt commented:
> I discovered a case where a bool value was set with memcpy and got the value 5. Since the GCC compiler assumes that a bool value is only true or false, it optimized out the existing check that sets the value to 0 or 1 and then unfortunately a wrong value is sent. This could lead to an error in the backend when this lwm2m message was parsed.
> 
> Since the behaviour changes with different compiler options (-O0 works), I suggest to fix this. Either by changing the parameter type or by removing the existing bool-to-0/1 conversion so that the outcome is always the same.

This feels fragile, I guess the same could be said for any function accepting a `bool` argument? IMO, the calling code is already "faulty", and maybe we should try to find the root, and fix it there with other compiler checks.

sarchey commented:
> > I discovered a case where a bool value was set with memcpy and got the value 5. Since the GCC compiler assumes that a bool value is only true or false, it optimized out the existing check that sets the value to 0 or 1 and then unfortunately a wrong value is sent. This could lead to an error in the backend when this lwm2m message was parsed.
> > Since the behaviour changes with different compiler options (-O0 works), I suggest to fix this. Either by changing the parameter type or by removing the existing bool-to-0/1 conversion so that the outcome is always the same.
> 
> This feels fragile, I guess the same could be said for any function accepting a `bool` argument? IMO, the calling code is already "faulty", and maybe we should try to find the root, and fix it there with other compiler checks.

Yes, that's true, but normally in C it doesn't matter if a bool value is 5 or 1, it's still true. In this case though, if a value outside of 0/1 is sent instead of a "real" bool value, the lwm2m/cbor parsing might get an error on the backend side. I think this is why there is already an existing check/conversion for this?

pdgendt commented:
> Yes, that's true, but normally in C it doesn't matter if a bool value is 5 or 1, it's still true. In this case though, if a value outside of 0/1 is sent instead of a "real" bool value, the lwm2m/cbor parsing might get an error on the backend side. I think this is why there is already an existing check/conversion for this?

I think the conversion was mostly done to guarantee the byte size to pass to the engine. The `value != 0` already strikes me as odd, honestly.

Can you share how the calling code is structured? A snippet with the `memcpy`?

rlubos commented:
The compliance errors need addressing:
>  LONG_LINE_COMMENT: line length of 114 exceeds 100 columns
File:subsys/net/lib/lwm2m/lwm2m_registry.c
Line:783
 BLOCK_COMMENT_STYLE: Block comments use a trailing */ on a separate line
File:subsys/net/lib/lwm2m/lwm2m_registry.c
Line:784

gramsay0 commented:
> This feels fragile, I guess the same could be said for any function accepting a `bool` argument? IMO, the calling code is already "faulty", and maybe we should try to find the root, and fix it there with other compiler checks.

I agree with @pdgendt. The bug is where a value other than 0 or 1 got into a boolean variable, rather than with this code. It would have to be memcpy'd or type punned from something that wasn't already a bool to bypass the implicit conversion to 0 or 1

yashi commented:
> **6.3.1.2 Boolean type**
> When any scalar value is converted to _Bool , the result is 0 if the value compares equal to 0;
> otherwise, the result is 1.

When you `memcpy()` to the location of a bool, you are killing (overwriting) your variable by yourself. C doesn't stop you to do that.

If you want to catch it, try `-fsanitize=bool`:

```c
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

uint8_t set_bool(bool value) {
    uint8_t ret = value != 0 ? 1 : 0;
    return ret;
}

int main(void)
{
	bool val;
	uint8_t u8 = 5;

	memcpy(&val, &u8, 1);

	(void)set_bool(val);

	return 0;
}
```

```
$ gcc -Wall -Wextra a.c -fsanitize=bool && ./a.out
a.c:17:8: runtime error: load of value 5, which is not a valid value for type '_Bool'
```

sarchey commented:
The original bug is in a library I used and I get the value via a CAN bus. Since I'm just parsing the data and copying it into a struct, I don't think `-fsanitize=bool` would help in my case.

However, we had a problem on the backend side which was quite hard to find as we mostly develop and test with CONFIG_DEBUG_OPTIMIZATIONS. Suddenly there was a problem with our daily bundle (CONFIG_SIZE_OPTIMIZATIONS).

I decided to make an upstream so that other people don`t have this problem too. But not everyone seems to like the change on the ‘conversion’, so my last suggestion is to remove the whole conversion or at least only make it active when `sizeof(bool) > 1`, which might only happen with a few compilers.

But I don't think it's up to me to remove this conversion, which is why I'm closing this PR for now.
Thanks to everyone for the valuable feedback though!
